runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/ClockSkewInterceptor.kt

```diff
@@ -0,0 +1,117 @@
+         * @return true if the service's response represents a definite clock skew error
+         * OR a *possible* clock skew error AND the skew exists. false otherwise.
+         * @param responseCodeDescription the server's response code description
+         * @param serverTime the server's time
+         */
+        internal fun Instant.isSkewed(serverTime: Instant, responseCodeDescription: String?): Boolean =
+            responseCodeDescription?.let {
+                CLOCK_SKEW_ERROR_CODES.contains(it) || (POSSIBLE_CLOCK_SKEW_ERROR_CODES.contains(it) && until(serverTime).absoluteValue >= CLOCK_SKEW_THRESHOLD)
+            } ?: false
+
+        // Errors definitely caused by clock skew
+        private val CLOCK_SKEW_ERROR_CODES = listOf(
+            "RequestTimeTooSkewed",
+            "RequestExpired",
+            "RequestInTheFuture",
+        )
+
+        // Errors possibly caused by clock skew
+        private val POSSIBLE_CLOCK_SKEW_ERROR_CODES = listOf(
+            "PriorRequestNotComplete",
+            "RequestTimeout",
+            "RequestTimeoutException",
+            "InternalError",
+        )
+    }
+
+    // Clock skew to be applied to all requests
+    private val _currentSkew: AtomicRef<Duration?> = atomic(null)
+
+    /**
+     * Apply the previously-computed skew, if it's set, to the execution context before signing
+     */
+    public override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        val logger = coroutineContext.logger<ClockSkewInterceptor>()
+
+        val skew = _currentSkew.value
+        skew?.let {
+            logger.info { "applying clock skew $skew to client" }
+            context.executionContext[HttpOperationContext.ClockSkew] = skew
+        }
+
+        context.executionContext[HttpOperationContext.ClockSkewApproximateSigningTime] = Instant.now()
+
+        return context.protocolRequest
+    }
+
+    /**
+     * After receiving a response, check if the client clock is skewed and apply a correction if necessary.
+     */
+    public override suspend fun modifyBeforeAttemptCompletion(context: ResponseInterceptorContext<Any, Any, HttpRequest, HttpResponse?>): Result<Any> {
+        val logger = coroutineContext.logger<ClockSkewInterceptor>()
+
+        val serverTime = context.protocolResponse?.header("Date")?.let {
+            Instant.fromRfc5322(it)
+        } ?: run {
+            logger.debug { "service did not return \"Date\" header, skipping skew calculation" }
+            return context.response
+        }
+
+        val clientTime = context.protocolRequest.headers["Date"]?.let {
+            Instant.fromRfc5322(it)
+        } ?: context.protocolRequest.headers["x-amz-date"]?.let {
+            Instant.fromIso8601(it)
+        } ?: context.executionContext[HttpOperationContext.ClockSkewApproximateSigningTime]
+
+        if (clientTime.isSkewed(serverTime, context.protocolResponse?.status?.description)) {
```

Comment:
**correctness**: This is using the `HttpStatusCode` description which is not the same as the the AWS error code that the `isSkewed` implementation is looking for. The change to predicate off exceptions probably means we need to move this interceptor into `aws-sdk-kotlin/aws-runtime/aws-http` package instead of `smithy-kotlin` since this is now AWS specific.